Dispatcher/UI/Auth/AuthViewController.swift

```diff
@@ -0,0 +1,231 @@
+        startLoadingScreen()
+        
+        let email = emailFormView.textField.text, password = passwordFormView.textField.text
+        var legitEmail = true, legitPassword = true
+        if !viewModel.isValidEmailAddress(email: email ?? "") {
+            emailFormView.displayWarning()
+            legitEmail = false
+        }
+        if password == nil {
+            passwordFormView.displayWarning()
+        }
+
+        if legitEmail && legitPassword {
+            viewModel.logUserToApp(email: email!, password: password!) { error in
+                if let error = error {
+                    print(error)
+                } else {
+                    self.clearAllUIElements()
+                    self.navigateIntoApp()
+                }
+            }
+        }
+        stopLoadingScreen()
+    }
+    
+    func navigateIntoApp() {
+        let homepage = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: "MyTabBarController")
+        self.present(homepage, animated: true, completion: nil)
+    }
+    
+    func startLoadingScreen() {
+        DispatchQueue.main.async {
+            self.loadingView.isHidden = false
+            self.loadingView.loadIndicator.startAnimating()
+        }
+    }
+    
+    func stopLoadingScreen() {
+        DispatchQueue.main.async {
+            self.loadingView.loadIndicator.stopAnimating()
+            self.loadingView.isHidden = true
+        }
+    }
+}
+
+
+// MARK: - FormInputViewDelegate
+extension AuthViewController: FormInputViewDelegate {
+    
+    func textFieldDidChange(textFieldId: String, currentText: String?) {
+        
+        if currentPageType == .signup {
+            if textFieldId == "email" {
+                if currentText != nil && currentText!.count > 5 {
+                    if !viewModel.isValidEmailAddress(email: currentText!) {
+                        emailFormView.displayWarning()
+                    } else {
+                        emailFormView.hideWarning()
+                    }
+                }
+            } else if textFieldId == "password" {
+                let isPasswordStrong = viewModel.isStrongPassword(password: currentText ?? "")
+                if !isPasswordStrong {
+                    passwordFormView.displayWarning()
+                } else {
+                    passwordFormView.hideWarning()
+                }
+            } else {
+                if currentText != nil && currentText!.count > 5 {
+                    if currentText != passwordFormView.textField.text {
+                        reenterPasswordFormView.displayWarning()
+                    } else {
+                        reenterPasswordFormView.hideWarning()
+                    }
+                }
+            }
+        }
+    }
+}
+
+
+
+// MARK: - MainActionButtonDelegate
+extension AuthViewController: MainActionButtonDelegate {
+    
+    func actionButtonDidPress(btnText: String) {
+
+        switch btnText {
+        case "LOGIN":
```

Comment:
Please set those raw text to static vars since you're using them also at other places ("LOGIN", "SIGNUP"...).
If you'll change this text you'll need to search where you use it and can miss places or get typo for one place.